lib/spack/spack/rewiring.py

```diff
@@ -0,0 +1,104 @@
+            raise RuntimeError('Pure spec package was not installed')
+        if spec.build_spec is not spec:
+            explicit = spec is spliced_spec
+            rewire_node(spec, explicit)
+
+
+def rewire_node(spec, explicit):
+    """This function rewires a single node, worrying only about references to
+    its subgraph. Binaries, text, and links are all changed in accordance with
+    the splice. The resulting package is then 'installed.'"""
+    tempdir = tempfile.mkdtemp()
+    # copy anything installed to a temporary directory
+    shutil.copytree(spec.build_spec.prefix,
+                    os.path.join(tempdir, spec.dag_hash()))
+
+    spack.hooks.pre_install(spec)
+    # compute prefix-to-prefix for every node from the build spec to the spliced
+    # spec
+    prefix_to_prefix = OrderedDict({spec.build_spec.prefix: spec.prefix})
+    for build_dep in spec.build_spec.traverse(root=False):
+        prefix_to_prefix[build_dep.prefix] = spec[build_dep.name].prefix
+
+    manifest = bindist.get_buildfile_manifest(spec.build_spec)
+    platform = spack.platforms.by_name(spec.platform)
+    if manifest.get('text_to_relocate'):
+        text_to_relocate = [os.path.join(tempdir, spec.dag_hash(), rel_path)
+                            for rel_path in manifest.get('text_to_relocate')]
+        relocate.relocate_text(files=text_to_relocate,
+                               prefixes=prefix_to_prefix)
+    if manifest.get('binary_to_relocate'):
+        bins_to_relocate = [os.path.join(tempdir, spec.dag_hash(), rel_path)
+                            for rel_path in manifest.get('binary_to_relocate')]
+        if 'macho' in platform.binary_formats:
+            relocate.relocate_macho_binaries(bins_to_relocate,
+                                             str(spack.store.layout.root),
+                                             str(spack.store.layout.root),
+                                             prefix_to_prefix,
+                                             False,
+                                             spec.build_spec.prefix,
+                                             spec.prefix)
+        if 'elf' in platform.binary_formats:
+            relocate.relocate_elf_binaries(bins_to_relocate,
+                                           str(spack.store.layout.root),
+                                           str(spack.store.layout.root),
+                                           prefix_to_prefix,
+                                           False,
+                                           spec.build_spec.prefix,
+                                           spec.prefix)
+        relocate.relocate_text_bin(binaries=bins_to_relocate,
+                                   prefixes=prefix_to_prefix)
+    # copy package into place (shutil.copytree)
+    shutil.copytree(os.path.join(tempdir, spec.dag_hash()), spec.prefix,
+                    ignore=shutil.ignore_patterns('.spack/spec.json'))
```

Comment:
This is going to get angry if the prefix already exists - if it does and has the same hash I think we already have the spec and don't need to do again?

```suggestion
    if not os.path.exists(spec.prefix):
        shutil.copytree(os.path.join(tempdir, spec.dag_hash()), spec.prefix,
                        ignore=shutil.ignore_patterns('.spack/spec.json'))
```
but if it's the case that the relocations are different (and the hash is the same) then you'd instead want to do:

This is going to get angry if the prefix already exists - if it does and has the same hash I think we already have the spec and don't need to do again?
```suggestion
    if os.path.exists(spec.prefix):
        shutil.rmtree(spec.prefix)
    shutil.copytree(os.path.join(tempdir, spec.dag_hash()), spec.prefix,
                    ignore=shutil.ignore_patterns('.spack/spec.json'))
```

Reply:
If the relocations are different the hash has to be different too -- the hash contains all of the information on all of the splices.

Reply:
Yup sorry for not resolving this earlier. We wound up fixing the issue with the check to install missing specs.